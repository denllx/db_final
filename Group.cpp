#include "Group.h"
#include <memory>
#include <vector>



int Group::size() const {
	return records.size();
}

shared_ptr<Record> Group::get_record(int id) const{
	return records[id];
}

void Group::addRecord(shared_ptr<Record> r) {
	records.push_back(r);
}

vector<Group> get_group(const vector<Record>& records,int keyid) {
	vector<Group> groups;
	int first = 0, tmp = 1,len=records.size(),ngroup=0;
	if (records.size() > 0) {
		while (tmp < len) {
			groups.push_back(Group());
			groups[ngroup].addRecord(make_shared<Record>(records[first]));
			tmp = first + 1;
			while (tmp < len 
				&& *(records[tmp].get_field(keyid)) == *(records[first].get_field(keyid))) {
				groups[ngroup].addRecord(make_shared<Record>( records[tmp++]));
			}
			ngroup++;
			first = tmp;
		}
	}
	return groups;
}