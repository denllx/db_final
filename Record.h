#pragma once
#include <vector>
#include <memory>
#include "Value.h"
using namespace std;

class Record{
    vector<shared_ptr<Value> > field;
public:
    Record(const vector<shared_ptr<Value> >&);
    shared_ptr<Value> get_field(int attr_id) const; /* 0-based */
    void update(int id, shared_ptr<Value>);
    int size() const;
};

class RecordComparator {
	int keyid;
public:
	RecordComparator(int key):keyid(key){}
	bool operator()(const Record&, const Record&) const;
};