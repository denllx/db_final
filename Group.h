#pragma once
#include <vector>
#include <memory>
#include "Record.h"
using namespace std;

class Group {
	vector<shared_ptr<Record> > records;
public:
	int size() const;
	shared_ptr<Record> get_record(int record_id) const;

	void addRecord(shared_ptr<Record> r);
};

vector<Group> get_group(const vector<Record>& records,int);
