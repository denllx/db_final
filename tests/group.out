stu_name
a
b
c
